backend/dal/Services/Concrete/ProjectReportService.cs

```diff
@@ -0,0 +1,258 @@
+        /// <exception cref="ArgumentNullException">Argument 'reportId' is required.</exception>
+        /// <exception cref="KeyNotFoundException">Project report does not exist in the datasource.</exception>
+        /// <exception cref="NotAuthorizedException">User does not have permission to view snapshots.</exception>
+        /// <returns></returns>
+        public IEnumerable<ProjectSnapshot> GetSnapshots(ProjectReport report)
+        {
+            this.User.ThrowIfNotAuthorized(Permissions.AdminProjects);
+
+            var snapshots = this.Context.ProjectSnapshots
+                .Include(s => s.Project).ThenInclude(p => p.Agency)
+                .Include(s => s.Project).ThenInclude(p => p.Status)
+                .Where(s => s.SnapshotOn == report.From || s.SnapshotOn == report.To)
+                .ToArray();
+
+            foreach(ProjectSnapshot snapshot in snapshots)
+            {
+                var previousSnapshot = snapshots.Where(s => s.ProjectId == snapshot.ProjectId && s.Id != snapshot.Id).FirstOrDefault();
+                snapshot.BaselineIntegrity = (snapshot?.NetProceeds ?? 0) - (previousSnapshot?.NetProceeds ?? 0);
+            }
+
+            return snapshots;
+        }
+
+        /// <summary>
+        /// Get all project snapshots matching either the 'To' or 'From' dates of the passed report.
+        /// </summary>
+        /// <param name="reportId"></param>
+        /// <exception cref="ArgumentNullException">Argument 'reportId' is required.</exception>
+        /// <exception cref="KeyNotFoundException">Project report does not exist in the datasource.</exception>
+        /// <exception cref="NotAuthorizedException">User does not have permission to view snapshots.</exception>
+        /// <returns></returns>
+        public IEnumerable<ProjectSnapshot> GetSnapshots(int reportId)
+        {
+            this.User.ThrowIfNotAuthorized(Permissions.AdminProjects);
+
+            var report = this.Context.ProjectReports
+                .FirstOrDefault(p => p.Id == reportId) ?? throw new KeyNotFoundException();
+
+            return GetSnapshots(report);
+        }
+
+        /// <summary>
+        /// Add the specified report to the datasource.
+        /// The 'From' date is set to the 'To' date of the most recent report, if one exists.
+        /// For every SPL project in the system, a new Snapshot is generated using current project data and a 'snapshotOn' time equal to the newly created reports 'To' date.
+        /// </summary>
+        /// <param name="project"></param>
+        /// <exception cref="ArgumentNullException">Argument 'report' is required.</exception>
+        /// <exception cref="NotAuthorizedException">User does not have permission to add report.</exception>
+        /// <returns></returns>
+        public ProjectReport Add(ProjectReport report)
+        {
+            if (report.To == null)
+            {
+                report.To = DateTime.UtcNow;
```

Comment:
Based on requested changes to the `GenerateSnapshots(...)` function this would be removed and you would determine the value from the results of the function.